File: src/test/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/integrasjoner/kafka/KafkaServiceIntegrasjonTest.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.kafka;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.module.paramnames.ParameterNamesModule;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.SykefraværsstatistikkLocalTestApplication;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.SykefraværFlereKvartalerForEksport;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.kafka.dto.KafkaStatistikkKategoriTopicValue;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.kafka.dto.KafkaTopicValue;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.Statistikkategori;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefravar.SykefraværMedKategori;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.UmaskertSykefraværForEttKvartal;
import no.nav.security.token.support.spring.test.EnableMockOAuth2Server;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
import org.springframework.kafka.listener.ContainerProperties;
import org.springframework.kafka.listener.KafkaMessageListenerContainer;
import org.springframework.kafka.listener.MessageListener;
import org.springframework.kafka.test.EmbeddedKafkaBroker;
import org.springframework.kafka.test.context.EmbeddedKafka;
import org.springframework.kafka.test.utils.ContainerTestUtils;
import org.springframework.kafka.test.utils.KafkaTestUtils;
import org.springframework.test.context.ActiveProfiles;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;

import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.__2019_3;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.__2019_4;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.__2020_1;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.__2020_2;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.assertKafkaStatistikkKategoriTopicValueEquals;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.assertKafkaTopicValueEquals;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.getKafkaTopicValueAsJsonString;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.landSykefravær;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.næring5SifferSykefravær;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.næringSykefravær;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.sektorSykefravær;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.virksomhetSykefravær;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

@ActiveProfiles("mvc-test")
@SpringBootTest(
        classes = SykefraværsstatistikkLocalTestApplication.class,
        webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT
)
@EnableMockOAuth2Server
@EmbeddedKafka(
        controlledShutdown = true,
        topics = {"arbeidsgiver.sykefravarsstatistikk-v1", "arbeidsgiver.sykefravarsstatistikk-land-v1"},
        brokerProperties = {"listeners=PLAINTEXT://localhost:9092", "port=9092"}
)
public class KafkaServiceIntegrasjonTest {

    @Autowired
    private KafkaService kafkaService;
    @Autowired
    private EmbeddedKafkaBroker embeddedKafkaBroker;

    private static String[] TOPIC_NAME = {
            "arbeidsgiver.sykefravarsstatistikk-v1",
            "arbeidsgiver.sykefravarsstatistikk-land-v1"
    };
    private KafkaMessageListenerContainer<String, String> container;
    private BlockingQueue<ConsumerRecord<String, String>> consumerRecords;
    private final static ObjectMapper objectMapper = new ObjectMapper();


    @BeforeEach
    public void setUp() {
        consumerRecords = new LinkedBlockingQueue<>();

        ContainerProperties containerProperties = new ContainerProperties(TOPIC_NAME);
        Map<String, Object> consumerProperties =
                KafkaTestUtils.consumerProps(
                        "consumer",
                        "false",
                        embeddedKafkaBroker
                );
        consumerProperties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
        consumerProperties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
        consumerProperties.put(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, "60000");
        DefaultKafkaConsumerFactory<String, String> kafkaConsumerFactory =
                new DefaultKafkaConsumerFactory<>(consumerProperties);

        container = new KafkaMessageListenerContainer<>(kafkaConsumerFactory, containerProperties);
        container.setupMessageListener((MessageListener<String, String>) record -> {
            System.out.println("Listened message=" + record.toString());
            consumerRecords.add(record);
        });
    }


    @AfterEach
    public void tearDown() {
        container.stop();
    }

    @Test
    public void send__sender_en_KafkaTopicValue_til__riktig_topic() throws Exception {
        container.start();
        ContainerTestUtils.waitForAssignment(container, 4);

        kafkaService.send(
                __2020_2,
                virksomhetSykefravær,
                Arrays.asList(næring5SifferSykefravær),
                næringSykefravær,
                sektorSykefravær,
                landSykefravær
        );

        ConsumerRecord<String, String> message = consumerRecords.poll(10, TimeUnit.SECONDS);

        assertNotNull(message);
        assertEquals("{\"orgnr\":\"987654321\",\"kvartal\":2,\"årstall\":2020}", message.key());
        objectMapper.registerModule(new ParameterNamesModule(JsonCreator.Mode.PROPERTIES));

        assertKafkaTopicValueEquals(
                objectMapper.readValue(getKafkaTopicValueAsJsonString(), KafkaTopicValue.class),
                objectMapper.readValue(message.value(), KafkaTopicValue.class)
        );
    }

    @Test
    public void send_til_LAND_topic___sender_en_KafkaTopicValue_til__riktig_topic() throws Exception {
        SykefraværFlereKvartalerForEksport sykefraværFlereKvartaler =  new SykefraværFlereKvartalerForEksport(
                List.of(
                        new UmaskertSykefraværForEttKvartal(
                                __2020_2,
                                new BigDecimal(1100),
                                new BigDecimal(11000),
                                5
                        ),
                        new UmaskertSykefraværForEttKvartal(
                                __2020_1,
                                new BigDecimal(2200),
                                new BigDecimal(22000),
                                5
                        ),
                        new UmaskertSykefraværForEttKvartal(
                                __2019_4,
                                new BigDecimal(3300),
                                new BigDecimal(33000),
                                5
                        ),
                        new UmaskertSykefraværForEttKvartal(
                                __2019_3,
                                new BigDecimal(4400),
                                new BigDecimal(44000),
                                5
                        )
                )
        );

        KafkaStatistikkKategoriTopicValue expectedTopicValue = new KafkaStatistikkKategoriTopicValue(
                new SykefraværMedKategori(
                        Statistikkategori.LAND,
                        landSykefravær.getKode(),
                        landSykefravær.getÅrstallOgKvartal(),
                        landSykefravær.getTapteDagsverk(),
                        landSykefravær.getMuligeDagsverk(),
                        landSykefravær.getAntallPersoner()
                ),
                sykefraværFlereKvartaler
        );

        container.start();
        ContainerTestUtils.waitForAssignment(container, 4);

        kafkaService.sendTilStatistikkKategoriTopic(
                __2020_2,
                Statistikkategori.LAND,
                "NO",
                landSykefravær,
                sykefraværFlereKvartaler
        );

        ConsumerRecord<String, String> message = consumerRecords.poll(10, TimeUnit.SECONDS);

        assertNotNull(message);
        assertEquals("{\"kategori\":\"LAND\",\"kode\":\"NO\",\"kvartal\":2,\"årstall\":2020}", message.key());
        objectMapper.registerModule(new ParameterNamesModule(JsonCreator.Mode.PROPERTIES));
        String value = message.value();
        KafkaStatistikkKategoriTopicValue actual = objectMapper.readValue(value, KafkaStatistikkKategoriTopicValue.class);
        assertKafkaStatistikkKategoriTopicValueEquals(
                expectedTopicValue,
                actual
        );
    }
}

File: src/test/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/integrasjoner/kafka/KafkaServiceIntegrasjonTest.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.kafka

import net.javacrumbs.jsonunit.assertj.assertThatJson
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.SykefraværsstatistikkLocalTestApplication
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.SykefraværFlereKvartalerForEksport
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.bransjeprogram.ArbeidsmiljøportalenBransje
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.ÅrstallOgKvartal
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.infrastruktur.config.KafkaTopicNavn
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.infrastruktur.config.KafkaTopicNavn.Companion.toStringArray
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.kafka.dto.MetadataVirksomhetKafkamelding
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.kafka.dto.Sektor
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.Statistikkategori
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.UmaskertSykefraværForEttKvartal
import no.nav.security.token.support.spring.test.EnableMockOAuth2Server
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.common.serialization.StringDeserializer
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Import
import org.springframework.kafka.core.DefaultKafkaConsumerFactory
import org.springframework.kafka.listener.ContainerProperties
import org.springframework.kafka.listener.KafkaMessageListenerContainer
import org.springframework.kafka.listener.MessageListener
import org.springframework.kafka.test.EmbeddedKafkaBroker
import org.springframework.kafka.test.utils.ContainerTestUtils
import org.springframework.kafka.test.utils.KafkaTestUtils
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.ActiveProfiles
import java.math.BigDecimal
import java.util.concurrent.BlockingQueue
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit

@ActiveProfiles("mvc-test", "kafka-test")
@SpringBootTest(
    classes = [SykefraværsstatistikkLocalTestApplication::class],
    webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT
)
@DirtiesContext
@EnableMockOAuth2Server
@Import(EmbeddedKafkaBrokerConfig::class)
class KafkaServiceIntegrasjonTest {
    @Autowired
    private lateinit var kafkaService: KafkaService

    @Autowired
    private lateinit var embeddedKafkaBroker: EmbeddedKafkaBroker

    private lateinit var container: KafkaMessageListenerContainer<String, String>
    private lateinit var consumerRecords: BlockingQueue<ConsumerRecord<String, String>>
    private val dummyData = SykefraværFlereKvartalerForEksport(
        listOf(
            UmaskertSykefraværForEttKvartal(
                EksporteringServiceTestUtils.__2020_2, BigDecimal(1100), BigDecimal(11000), 5
            )
        )
    )

    @BeforeEach
    fun setUp() {
        consumerRecords = LinkedBlockingQueue()
        val containerProperties = ContainerProperties(*TOPIC_NAMES)
        val consumerProperties =
            KafkaTestUtils.consumerProps("consumer", "false", embeddedKafkaBroker)
        consumerProperties[ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG] =
            StringDeserializer::class.java
        consumerProperties[ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG] =
            StringDeserializer::class.java
        consumerProperties[ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG] = "60000"
        val kafkaConsumerFactory = DefaultKafkaConsumerFactory<String, String>(consumerProperties)
        container = KafkaMessageListenerContainer(kafkaConsumerFactory, containerProperties)
        container.setupMessageListener(
            MessageListener { record: ConsumerRecord<String, String> ->
                println("Listened message=$record")
                (consumerRecords as LinkedBlockingQueue<ConsumerRecord<String, String>>).add(record)
            } as MessageListener<String, String>)
        container.start()
        ContainerTestUtils.waitForAssignment(
            container, embeddedKafkaBroker.partitionsPerTopic * TOPIC_NAMES.size
        )
    }

    @AfterEach
    fun tearDown() {
        container.destroy()
    }

    @Test
    fun `send kafkamelding med metadata sender på riktig topic`() {
        kafkaService.send(
            MetadataVirksomhetKafkamelding(
                "999999999",
                ÅrstallOgKvartal(2023, 2),
                "86101",
                ArbeidsmiljøportalenBransje.SYKEHUS,
                Sektor.STATLIG
            ),
            KafkaTopicNavn.SYKEFRAVARSSTATISTIKK_METADATA_V1
        )
        val message = consumerRecords.poll(10, TimeUnit.SECONDS)
        assertThat(message?.topic()).isEqualTo(KafkaTopicNavn.SYKEFRAVARSSTATISTIKK_METADATA_V1.topic)
    }

    @Test
    fun `send kafkamelding med metadata sender riktig data`() {
        kafkaService.send(
            MetadataVirksomhetKafkamelding(
                "999999999",
                ÅrstallOgKvartal(2023, 2),
                "86101",
                ArbeidsmiljøportalenBransje.SYKEHUS,
                Sektor.STATLIG
            ),
            KafkaTopicNavn.SYKEFRAVARSSTATISTIKK_METADATA_V1
        )
        val message = consumerRecords.poll(10, TimeUnit.SECONDS)
        assertThatJson(message!!.value()) {
            isObject
            node("orgnr").isString.isEqualTo("999999999")
            node("arstall").isString.isEqualTo("2023")
            node("kvartal").isString.isEqualTo("2")
            node("bransje").isString.isEqualTo("SYKEHUS")
            node("naringskode").isString.isEqualTo("86101")
            node("sektor").isString.isEqualTo("STATLIG")
        }
    }

    @Test
    fun `send kafkamelding for alle kategorier sender melding til riktig topic`() {
        kafkaService.send(
            EksporteringServiceTestUtils.__2020_2,
            EksporteringServiceTestUtils.virksomhetSykefravær,
            listOf(EksporteringServiceTestUtils.næring5SifferSykefravær),
            EksporteringServiceTestUtils.næringSykefravær,
            EksporteringServiceTestUtils.sektorSykefravær,
            EksporteringServiceTestUtils.landSykefravær
        )
        val message = consumerRecords.poll(10, TimeUnit.SECONDS)
        Assertions.assertEquals(KafkaTopicNavn.SYKEFRAVARSSTATISTIKK_V1.topic, message!!.topic())
    }

    @Test
    fun `send kafkamelding for landkategori sender melding til riktig topic`() {
        kafkaService.sendTilStatistikkKategoriTopic(
            EksporteringServiceTestUtils.__2020_2,
            Statistikkategori.LAND,
            "NO",
            EksporteringServiceTestUtils.landSykefravær,
            dummyData
        )
        val message = consumerRecords.poll(10, TimeUnit.SECONDS)
        Assertions.assertEquals(
            KafkaTopicNavn.SYKEFRAVARSSTATISTIKK_LAND_V1.topic,
            message!!.topic()
        )
    }

    @Test
    fun `send for kategori næring sender til "sykefravarsstatistikk-næring-topic-v1"`() {
        kafkaService.sendTilStatistikkKategoriTopic(
            EksporteringServiceTestUtils.__2020_2,
            Statistikkategori.NÆRING,
            "11",
            EksporteringServiceTestUtils.næringSykefravær,
            dummyData
        )
        val message = consumerRecords.poll(10, TimeUnit.SECONDS)
        Assertions.assertEquals(
            KafkaTopicNavn.SYKEFRAVARSSTATISTIKK_NARING_V1.topic,
            message!!.topic()
        )
    }

    @Test
    fun send__forSektorKategori__senderTilSykefravarsstatistikkSektorV1Topic() {
        kafkaService.sendTilStatistikkKategoriTopic(
            EksporteringServiceTestUtils.__2020_2,
            Statistikkategori.SEKTOR,
            "11",
            EksporteringServiceTestUtils.næringSykefravær,
            dummyData
        )
        val message = consumerRecords.poll(10, TimeUnit.SECONDS)
        Assertions.assertEquals(
            KafkaTopicNavn.SYKEFRAVARSSTATISTIKK_SEKTOR_V1.topic,
            message!!.topic()
        )
    }

    @Test
    fun send__forVirksomhetKategori__senderTilSykefravarsstatistikkVirksomhetV1Topic() {
        kafkaService.sendTilStatistikkKategoriTopic(
            EksporteringServiceTestUtils.__2020_2,
            Statistikkategori.VIRKSOMHET,
            "11",
            EksporteringServiceTestUtils.virksomhetSykefraværMedKategori,
            dummyData
        )
        val message = consumerRecords.poll(10, TimeUnit.SECONDS)
        Assertions.assertEquals(
            KafkaTopicNavn.SYKEFRAVARSSTATISTIKK_VIRKSOMHET_V1.topic,
            message!!.topic()
        )
    }

    companion object {
        private val TOPIC_NAMES = toStringArray()
    }
}